src/core/ext/transport/chaotic_good/chaotic_good_transport.h

```diff
@@ -0,0 +1,113 @@
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_SRC_CORE_EXT_TRANSPORT_CHAOTIC_GOOD_CHAOTIC_GOOD_TRANSPORT_H
+#define GRPC_SRC_CORE_EXT_TRANSPORT_CHAOTIC_GOOD_CHAOTIC_GOOD_TRANSPORT_H
+
+#include <grpc/support/port_platform.h>
+
+#include "absl/random/random.h"
+
+#include "src/core/ext/transport/chaotic_good/frame.h"
+#include "src/core/ext/transport/chaotic_good/frame_header.h"
+#include "src/core/ext/transport/chttp2/transport/hpack_encoder.h"
+#include "src/core/lib/promise/if.h"
+#include "src/core/lib/promise/promise.h"
+#include "src/core/lib/promise/try_join.h"
+#include "src/core/lib/promise/try_seq.h"
+#include "src/core/lib/transport/promise_endpoint.h"
+
+namespace grpc_core {
+namespace chaotic_good {
+
+class ChaoticGoodTransport {
+ public:
+  ChaoticGoodTransport(std::unique_ptr<PromiseEndpoint> control_endpoint,
+                       std::unique_ptr<PromiseEndpoint> data_endpoint)
+      : control_endpoint_(std::move(control_endpoint)),
+        data_endpoint_(std::move(data_endpoint)) {}
+
+  auto WriteFrame(const FrameInterface& frame) {
+    auto buffers = frame.Serialize(&encoder_);
+    return TryJoin<absl::StatusOr>(
+        control_endpoint_->Write(std::move(buffers.control)),
+        data_endpoint_->Write(std::move(buffers.data)));
+  }
+
+  // Read frame header and payloads for control and data portions of one frame.
+  // Resolves to StatusOr<tuple<FrameHeader, BufferPair>>.
+  auto ReadFrameBytes() {
+    return TrySeq(
+        control_endpoint_->ReadSlice(FrameHeader::frame_header_size_),
+        [this](Slice read_buffer) {
+          auto frame_header =
+              FrameHeader::Parse(reinterpret_cast<const uint8_t*>(
+                  GRPC_SLICE_START_PTR(read_buffer.c_slice())));
+          // Read header and trailers from control endpoint.
+          // Read message padding and message from data endpoint.
+          return If(
+              frame_header.ok(),
+              [this, &frame_header] {
+                const uint32_t message_padding = std::exchange(
+                    last_message_padding_, frame_header->message_padding);
```

Comment:
Is the value of `last_message_padding_` used somewhere else?

Reply:
just here... std::exchange does a read/write operation on it